src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -2684,8 +2683,15 @@ exprt c_typecheck_baset::do_special_functions(
 
     typecheck_function_call_arguments(expr);
 
+    symbol_exprt deallocated =
+      lookup(CPROVER_PREFIX "deallocated").symbol_expr();
+    symbol_exprt dead = lookup(CPROVER_PREFIX "dead_object").symbol_expr();
```

Comment:
There's something wrong here. The connection to `deallocated` is very much a mechanism for implementing the predicate, and that shouldn't happen in the front-end.

Reply:
Would it help if we lowered the side-effecty function to a predicate in the built-in library?
The C implementation of the function can refer to the global state to call the built-in predicate.